src/proto/ctl/smd.proto

```diff
@@ -80,7 +80,7 @@ message SmdDevResp {
 		string uuid = 1; // UUID of blobstore
 		repeated int32 tgt_ids = 2; // VOS target IDs
 		string tr_addr = 4; // Transport address of blobstore
-		uint32 dev_state = 5; // NVMe device state
+		string dev_state = 5; // NVMe device state
```

Comment:
I don't understand why the dRPC messages need to be changed. Changing JSON makes sense, but all of our internal components can keep using the bitmasks, can't they? They should all have access to the DAOS header files with the translation information.

It looks like the dRPC stuff is all that touches C code. If that's the case, the string translation functions could be implemented more simply in Go.

Reply:
There has been lots of discussion on this, see https://daosio.atlassian.net/browse/DAOS-9575 and https://daosio.atlassian.net/browse/DAOS-9559 for details. I recently changed the pb state to a bitset and was disinclined to change back to a string so I proposed an update to just the JSON dmg output initially. @mjmac wanted us to be consistent with the JSON marshalling and unmarshalling of native struct including the protobuf path so that's what I implemented. Maybe he can offer better rational.

Reply:
After reviewing the tickets I'm still a bit confused. I think limiting it to a user output level (as you suggested) would make sense, just tacking on Marshal/UnmarshalJSON methods to the type may do the trick. I'm a bit leery of changing existing fields in the protobuf messages. @mjmac if you'd like to weigh in, I'm interested in hearing your thoughts.

Reply:
For context, here is what I said when discussing this work:
> the simplest way to think about this is: imagine you are a developer working with a closed-source product which has tools that emit JSON as an option. you want to integrate the product into other stuff. would you rather have symbolic representations that you can readily understand just by reading the output, or would you be ok with having numbers that you have to go look up in documentation and then encode in a map which may break as things change over time? 

> I still don’t really see why it’s a problem to use the string as an intermediate representation even for gRPC/dRPC though. The serialization stuff should be hidden from most daos developers, so not sure why anyone would care. 

> We don’t currently support direct connections from non-control clients to the gRPC endpoints, and likely never will, but if we were developing a more traditional set of network service endpoints we’d be having the same problem with representing the state in a portable manner. 

So, the point I was trying to make is that we need to be careful in how we think about the protobuf messages and JSON output. It's true that we currently can safely assume that both sides of a dRPC/gRPC call will have access to headers in order to resolve numeric message values into symbolic values that we can use in our code. It's also true that this is unlikely to change (i.e. we won't have to deal with "raw" gRPC clients connecting to our endpoints).

The main thing that I was trying to accomplish here was to make sure that JSON consumers do not need to maintain their own maps or functions in order to convert 
 numeric values into symbolic values for use in external codebases which may not have access to DAOS internals, because that is going to be an ongoing maintenance problem and source of bug reports, particularly as more external users integrate DAOS into their own tooling and infrastructure.

I should acknowledge that there is some potential tension between opposing goals here, though. This is all somewhat subject to judgement calls, as we also want to avoid forcing JSON consumers to maintain complex logic for transforming raw values into user-displayable output when possible.

So with all of that having been said, I'm still reading through the latest revision of this PR to understand what has been implemented. I think that Tom has tried to thread the needle on this and I think it's reasonably close, but I'm a little concerned about ongoing maintenance of the `nvme_*2*()` helpers, at least as they're currently written. If we don't expect them to change much from where they are now, then maybe they're fine.

Does that help, @kjacque?

Reply:
Thanks for the details, Mike. It sounds like we're in agreement on the JSON side, that at the very least we don't want an opaque number for the user to decipher. I mentioned above that perhaps they could be a set of booleans, since the string itself is essentially a representation of that, which could help create some clarity for users scripting to the output. I'm not too picky about that though.

> It's true that we currently can safely assume that both sides of a dRPC/gRPC call will have access to headers in order to resolve numeric message values into symbolic values that we can use in our code. It's also true that this is unlikely to change (i.e. we won't have to deal with "raw" gRPC clients connecting to our endpoints).

That's why I don't see any harm in keeping the current bitmap format internally in these messages, since that's the format the engines are using in the C code. I don't see the usefulness of having engines convert to and from strings that the control plane deciphers on the other side anyway.

Reply:
Yes I've definitely been trying to "thread the needle", @sydneyvanda and @NiuYawei have been specific about the state string mappings and it's not exactly linear so I think that's the best I can do for the moment in order to take requirements into account. Are we good to go then for the moment or are there more changes required? I've added test coverage and error handling for unknown string cases in the most recent commit.

Reply:
The strings themselves are OK with me. It's just the protobuf changes that are bothering me at this point. But if we're set on them, I suppose I'll disagree and commit. :)

Reply:
> I still don’t really see why it’s a problem to use the string as an intermediate representation even for gRPC/dRPC though. The serialization stuff should be hidden from most daos developers, so not sure why anyone would care.

My concern was that it's not convenient to use string accurately representing a bitset (we have to define many obscure strings to accurately represent every combination of the state bits). For example, what if user requires a '--verbose' option for 'dmg list devices' to display the accurate device state? (I suppose the device state could be displayed as a bitmap to user then, some thing like "in-use:yes, plugged:yes, faulty:no, identify:no"), then I think io engine has to reply the intact bitset to control plane to ensure no information loss during transfer.

I was thinking that the io engine is only responsible for transferring accurate dev state to control plane, and control plane is responsible for displaying it to end user in whatever format, now I see that we also want the response in some human-readable JSON format.

Anyway, I think current state strings ("NEW", "EVICTED", "UNPLUGGED", "NORMAL") should work well if user doesn't require 'verbose" output on device state, so I'm good with both ways. I just try to clarify why I want to use bitset instead of strings at the very beginning.